Feature/form

TemVin opened the pull request.

TemVin committed:
Added application logic, implemented a mechanism to fill out the document when filling out the form

TemVin committed:
added form validation, app styling has been added

TemVin committed:
Added email validation, fixed appearance app

TemVin committed:
add validate

TemVin committed:
add event onblur, cleaned code

TemVin committed:
corrected code

TemVin committed:
add background image

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,152 @@
+import React, { PureComponent } from 'react';
+
```
ненужна пустая строка

TemVin replied:
Исправил

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,152 @@
+import CSSModules from 'react-css-modules';
+
+import styles from './BlankApp.scss';
+import Form from './Form';
```
Файлы должны подключаться так:
- внешние библиотеки 
- пустая строка
- компноненты из папок проекта (_components, _utils и т.п.) - если они нужны
- пустая строка
- локальные компоненты (./component)
- пустая строка 
- import styles from './BlankApp.scss';

TemVin replied:
Запомнил и исправил

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,152 @@
+            email={email}
+            changeInput={this.updateInput}
+            changeTextareaStory={this.changeTextareaStory}
+            isValidateForm={formValid}
```
и в итоге ты это поле даже не используешь в форм

TemVin replied:
Было раньше для кнопки, убрал

TemVin committed:
corrected code

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,152 @@
+
+  updateInput = ({ target: { name, value } }) => {
+    this.setState({ [name]: value },
+      () => { this.validateField(name, value); },
```
а зачем писать это именно в колбек?
почему нельзя просто написать строчкой ниже отдельно?

TemVin replied:
исправил

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,152 @@
+      lastname: '',
+      date: '',
+    },
+    emailValid: false,
```
эти поля лишние, потому что если например formErrors.email пуста строка - это false и наоборот

TemVin replied:
убрал

TemVin committed:
corrected code css

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,134 @@
+
+class BlankApp extends PureComponent {
+  state = {
+    name: '',
```
У тебя есть `name`, такое же поле есть в `formErrors `
можно например, сделать name объектом у которого будет поле vallue и error
`name: {
  value: '',
  error: '',
}`
или вообще написать массив объектов fields и потом по нему проходиться и отрисовывать все поля, но это сложнее.
В любом случае нужно связать value поля с его ошибкой, а сейчас они у тебя не то что лежат отдельно, но и имеею разную структуру данных поля как строки, а ошибки почему то собраны в один объект `formErrors `

TemVin replied:
Написал массив объетов fields

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,134 @@
+    dateValid: false,
+  };
+
+  changeTextareaStory = (textareaStory) => {
```
ты же изменяешь поле story, тогда пиши `changeStory`
и вместо `textareaStory` просто value или на крайняк `story`

TemVin replied:
Хорошо!

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,134 @@
+      lastnameValid,
+      dateValid,
+    } = this.state;
+    switch (fieldName) {
```
перед switch пробел нужен для читабельности

TemVin replied:
окей, исправил, буду ставить.

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,134 @@
+    } = this.state;
+    switch (fieldName) {
+      case 'name':
+        nameValid = value.trim().length === 0;
```
эта строка и 53 идентичны, выноси в отдельный метод или утилиту

TemVin replied:
вынес в отдельный метод

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,134 @@
+    switch (fieldName) {
+      case 'name':
+        nameValid = value.trim().length === 0;
+        formErrors.name = nameValid ? ' Поле "имя" должно быть заполнено' : '';
```
ты изменяешь состояние напрямую, без setState это грубейшая ошибка
здесь и далее

TemVin replied:
Да.. изменил во всех местах

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,134 @@
+        break;
+      case 'lastname':
+        lastnameValid = value.trim().length === 0;
+        formErrors.lastname = lastnameValid ? 'Поле "фамилия" должно быть заполнено' : '';
```
опять таки эта строчка идентична 50, только меняется название поля

TemVin replied:
вынес в отдельный метод

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    this.setState({
+      formErrors,
```
и смысл этого setState если ты уже всё и так поменял напрямую?
нужно менять эти поля на новые значения

TemVin replied:
убрал

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,134 @@
+
+  render() {
+    const {
+      name,
```
напрашивается массив, столько полей руками отрисовывать, слишком долго

TemVin replied:
теперь массив

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -0,0 +1,134 @@
+      styleName,
+    } = this.state;
+    return (
+      <div className={styleName} styleName="root">
```
`className={styleName}` - это работает?

TemVin replied:
поправил на это className={className}

konovalho commented on src/components/BlankApp/BlankApp.scss:
```diff
@@ -0,0 +1,23 @@
+  font-weight: 300;
+  font-size: 35px;
+  text-align: center;
+  margin-bottom: 30px;
```
https://react-best-practice-prototype.gitbook.io/project/#svoistva-silnee-vliyayushie-na-dokument-idut-pervymi

TemVin replied:
Исправил 
  margin-bottom: 30px;
  text-align: center;
  font-weight: 300;
  font-size: 35px;
  color: $color-white;

konovalho commented on src/components/BlankApp/Document/Document.jsx:
```diff
@@ -0,0 +1,40 @@
+import React, { PureComponent } from 'react';
+import CSSModules from 'react-css-modules';
+
```
пустая строка не нужна, перед styles наоборот нужна

TemVin replied:
исправил
import React, { PureComponent } from 'react';
import CSSModules from 'react-css-modules';
import PropTypes from 'prop-types';

import styles from './Document.scss';

konovalho commented on src/components/BlankApp/Document/Document.scss:
```diff
@@ -0,0 +1,37 @@
+
+.root {
+  padding: 15px;
+  background-color: rgba(10, 80, 100, 0.9);
```
цвет вынести в variables

TemVin replied:
вынес, назвал dark-cyan

konovalho commented on src/components/BlankApp/Form/FieldForm/FieldForm.jsx:
```diff
@@ -0,0 +1,68 @@
+      handleChangeInput,
+      placeholder,
+      formErrors,
+    } = this.props;
```
пустая строка нужна

TemVin replied:
теперь есть

konovalho commented on src/components/BlankApp/Form/FieldForm/FieldForm.jsx:
```diff
@@ -0,0 +1,68 @@
+      placeholder,
+      formErrors,
+    } = this.props;
+    const hasError = this.checkClassName();
```
перед return всегда пустая строка

TemVin replied:
Запомнил

konovalho commented on src/components/BlankApp/Form/Form.jsx:
```diff
@@ -0,0 +1,94 @@
+          <h3 styleName="title"> Введите информацию о себе </h3>
+          <div styleName="formFlex">
+            <p styleName="description">*поля, обязательные для заполнения</p>
+            <FieldForm
```
4 раза `FieldForm ` почти с одинаковыми свойствами, просто напрашивается проход по массиву

TemVin replied:
сделал проход по массиву

ArtemiusL commented:
компонент BlankApp ты вряд ли будешь переиспользовать, это слишком уникальный компонент со своей логикой, его не нужно помещать в _components, он должен лежать в папке страницы на которой ты его подключаешь

konovalho commented on src/components/BlankApp/Form/FieldForm/FieldForm.jsx:
```diff
@@ -0,0 +1,68 @@
+@CSSModules(styles, { allowMultiple: true })
+class FieldForm extends PureComponent {
+  checkClassName = () => {
+    const { formErrors, fieldName } = this.props;
```
в `FieldForm ` не нужно передавать весь объект с ошибками, можно передать конкретную и не морочиться с `formErrors[fieldName]`

TemVin replied:
немного по-другому вышло, посмотри потом по возможности, пожалуйста

konovalho commented on src/components/BlankApp/Form/FieldForm/FieldForm.jsx:
```diff
@@ -0,0 +1,68 @@
+            value={data}
+            onChange={handleChangeInput}
+            placeholder={placeholder}
+            onBlur={handleChangeInput}
```
если onBlur обрабатывает `handleChangeInput `, тогда он вообще получается не нужен
либо делает отдельный метод для onBlur где будет своя логика, либо удаляя, потому что сейчас в нем нет смысла

TemVin replied:
сделал отдельный метод для onBlur

konovalho commented on src/components/BlankApp/Form/FieldForm/FieldForm.jsx:
```diff
@@ -0,0 +1,68 @@
+            onBlur={handleChangeInput}
+          />
+        </label>
+        { Object.keys(formErrors).map((name) => {
```
опять таки передавай сюда конкретную ошибку и не мучайся

TemVin replied:
Теперь передаю конкретную ошибку

konovalho commented on src/components/BlankApp/Form/FieldForm/FieldForm.scss:
```diff
@@ -0,0 +1,32 @@
+  width: 100%;
+  height: 100px;
+
+  label {
```
а что label не заслуживает отдельного стиля?

TemVin replied:
Заслуживает)) 
Исправил

konovalho commented on src/components/BlankApp/Form/FieldForm/FieldFormError/FieldFormError.jsx:
```diff
@@ -0,0 +1,24 @@
+  render() {
+    const {
+      error,
+    } = this.props;
```
пустая строка

TemVin replied:
исправил

TemVin committed:
changed data structure, validation. Corrected some notes

TemVin committed:
add color in variables, swapped properties in accordance with the rules, add new selector for label

TemVin committed:
add new methods for validation, fixed code(add necessary spaces and indents, removed unused methods)

TemVin committed:
transferred the BlankApp component to _pages and changed its name to App

konovalho commented on src/components/BlankApp/BlankApp.jsx:
```diff
@@ -1,127 +1,147 @@
+        error: '',
+      },
+    ],
+    story: {
```
Просто строка. Объект здесь ни к чему, если в нем одно значение

TemVin replied:
Исправил, сделал в проекте story строкой

konovalho commented on src/pages/HomePage/App/App.jsx:
```diff
@@ -0,0 +1,150 @@
+      (!(/([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i).test(value.trim())) ? 'Неправильный формат еmail' : '');
+  }
+
+  handleChangeFieldErr = (fieldName, value, textErr) => {
```
Если value ничего не меняет, то его не нужно передавать

TemVin replied:
Согласен. Первоначально думал, что стоит хранить значение) Но сейчас убедился, что она не играет никакой роли

konovalho commented on src/pages/HomePage/App/App.jsx:
```diff
@@ -1,12 +1,12 @@
 
-import Form from './Form';
-import Document from './Document';
+import Form from '_components/BlankApp/Form';
```
Артем в чем смысл тогда этих изменений, если ты всё равно оставил BlankApp в _componets?
я же написал, что его нужно перенести - это значит что он весь, со всеми элементами должен быть перенесен, а не только тело компонента.
То есть Form и Document должны лежать внутри твоего нынешнего App

TemVin replied:
переместил

konovalho commented on src/pages/HomePage/HomePage.jsx:
```diff
@@ -1,13 +1,16 @@
-import Users from './Users';
+import App from './App';
+
+import styles from '_styles/style.scss';
```
Смотри, сейчас у тебя одна страница - HomePage, а потом будет ArticlePage, AboutPage, всё что угодно, а ты в _styles/style.scss прописал глобальный стиль container, а если его для других страниц нужно будет поменять?
Поэтому для HomePage должен быть отдельный файл HomePage.scss здесь же где и Homepage сам лежит

TemVin replied:
внес изменения

konovalho commented on src/pages/HomePage/App/App.jsx:
```diff
@@ -0,0 +1,150 @@
+  }
+
+  isEmptyField = (fieldName, value) => {
+    this.handleChangeFieldErr(fieldName, value, (value.trim().length === 0) ? 'Поле должно быть заполнено!' : '');
```
в трех методах используется `this.handleChangeFieldErr`, можно тоже сократить.
+ все эти методы `isEmptyField `, `isDateValid `, `isEmailValid ` подразумевается что возвращают либо true либо false ( сюда по названию ) - так и делай

TemVin replied:
выполнено.
Теперь так 
case 'name':
        this.handleChangeFieldErr(fieldName, this.isEmptyField(value));
        break;

isEmptyField = (value) => {
    if (value.trim().length === 0) {
      return 'Поле должно быть заполнено!';
    }
    return '';
  }

konovalho replied:
можно использовать тернарный оператор.

isEmptyField = (value) => (value.trim().length === 0 ? 'Поле должно быть заполнено!' : ''); 
Здесь и в подобных методах

konovalho commented on src/components/BlankApp/Form/Form.jsx:
```diff
@@ -0,0 +1,104 @@
+  {
+    id: 1,
+    fieldName: 'name',
+    label: 'Введите имя*',
```
символ `*` - указывает, что поле обязательное к заполнению, он не является частью label, он должен подставляться в зависимости от условия

TemVin replied:
выполнил, прописал в константы определённых полей, которые обязательны для заполнения. Или нужно обязательно от условия?)

TemVin replied:
Изменил. Использовал встроенный If с логическим оператором &&

konovalho commented on src/components/BlankApp/Form/Form.jsx:
```diff
@@ -0,0 +1,104 @@
+class Form extends PureComponent {
+  handleChangeStory = (evt) => {
+    const { changeStory } = this.props;
+    changeStory(evt.target.value);
```
если деструктуризируется только одно свойство, лучше вообще это не делать и просто написать 
`this.props.changeStory(evt.target.value);` - так будет короче

TemVin replied:
Поправил

konovalho commented on src/components/BlankApp/Form/FieldForm/FieldForm.jsx:
```diff
@@ -0,0 +1,58 @@
+            onBlur={handleCkeckValidate}
+          />
+        </label>
+        <FieldFormError
```
Если ошибка есть - то отрисовываешь компонент, если нет - не отрисовываешь

TemVin replied:
сделано

konovalho replied:
Неправильно, нужно так:
`{error  && (<FieldFormError
          error={error}
        />)}`

TemVin replied:
заменил. сделал также с hasError
styleName={classnames('input', error && { hasError: true })}

TemVin committed:
created  HomePage.scss file and put there class container

TemVin committed:
changed the position of some files (moved components Form and Document to pages)
corrected comments and improved code(e.g. the field 'FieldFormError' is rendered only in case of an error)

konovalho commented on src/pages/HomePage/App/App.jsx:
```diff
@@ -0,0 +1,150 @@
+      case 'name':
+        this.handleChangeFieldErr(fieldName, this.isEmptyField(value));
+        break;
+      case 'lastname':
```
этот case можно объединить с 'name'

TemVin replied:
объединил)

konovalho commented on src/components/FieldForm/FieldForm.jsx:
```diff
@@ -9,23 +9,32 @@ import styles from './FieldForm.scss';
+      );
+    }
+    return false;
+  }
```
не забывай пустую строку между методами

TemVin replied:
да, надо быть внимательнее)

konovalho commented on src/pages/HomePage/App/Document/Document.jsx:
```diff
@@ -0,0 +1,45 @@
+import styles from './Document.scss';
+
+@CSSModules(styles, { allowMultiple: true })
+class Document extends PureComponent {
```
нужно переназвать этот компонент.
В браузере есть глобальный объект `document` и у нас `Document`, потенциально могут быть какие-нибудь проблемы

TemVin replied:
Заменил на 'Blank'

konovalho commented on src/pages/HomePage/App/Form/Form.jsx:
```diff
@@ -0,0 +1,107 @@
+
+    return constantFields.map((item) => {
+      const { fieldName } = item;
+      const currentFields = fields.filter(field => field.name === fieldName);
```
Давай всё-таки будем использовать `indexOf()` - так как он больше подходит для поиска именно конкретного элемента + нужно обязательно проверить найден ли элемент, если элемент не найден функция должна вернуть `null`

TemVin replied:
const currentFieldIndex = fields.map(field => field.name).indexOf(fieldName);
      if (currentFieldIndex === -1) {
        return null;
      }

TemVin committed:
changed name component 'Document' to 'Blank', add indexOf to Form, combined case 'name' and 'lastname'

konovalho commented on src/pages/HomePage/App/Form/Form.jsx:
```diff
@@ -50,7 +50,10 @@ class Form extends PureComponent {
+      if (currentFields === -1) {
+        return null;
+      }
 
```
currentFieldIndex = 
currentField = fields[currentFieldIndex];

TemVin replied:
исправил

TemVin committed:
add change
--currentField = fields[currentFieldIndex];

TemVin committed:
add change
-- added ternary operator for methods: isEmptyField, isDateValid, isEmailValid
